lighthouse-plugin-publisher-ads/audits/viewport-ad-density.js

```diff
@@ -64,22 +63,24 @@ class ViewportAdDensity extends Audit {
       return auditNotApplicable.NoVisibleSlots;
     }
 
-    const adArea = slots.reduce((sum, slot) =>
-      sum + boxViewableArea(slot.clientRect, viewport), 0);
-
-    const viewArea = viewport.innerWidth * viewport.innerHeight;
-
-    if (viewArea <= 0) {
+    if (viewport.innerHeight <= 0) {
       throw new Error(auditError.ViewportAreaZero);
     }
-    if (adArea > viewArea) {
-      throw new Error(auditError.AreaLargerThanViewport);
-    }
-    const adDensity = adArea / viewArea;
+
+    // We measure document length based on the bottom ad so that it isn't skewed
+    // by lazy loading.
+    const adsBottom =
+      Math.max(...slots.map((s) => s.clientRect.top + s.clientRect.height / 2));
+    const documentLength = adsBottom + viewport.innerHeight;
+
+    const adHeights =
+      slots.map((s) => s.clientRect.height).reduce((a, b) => a + b, 0);
```

Comment:
As discussed offline, I think this should instead be number of vertical pixels containing an ad.

Reply:
Yup, I was planning to do that.